### ISSUE TITLE
AAF Reader: File can no longer be parsed after #460

### ISSUE THREAD

**Issue by andrewmoore-nz** [PixarAnimationStudios/OpenTimelineIO#484]
Have just come across a file I'd been using for testing that was previously being parsed (mostly) correctly by OTIO but is now throwing an error:

**CannotComputeAvailableRangeError: No available_range set on media reference on clip**
```
WARNING:root:fat sector count missmatch difat: 236 header: 138
Traceback (most recent call last):
  File "/usr/local/bin/otiocat", line 11, in <module>
    load_entry_point('OpenTimelineIO==0.10.0.dev1', 'console_scripts', 'otiocat')()
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio/console/otiocat.py", line 132, in main
    read_adapter_arg_map
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio/console/otiocat.py", line 99, in _otio_compatible_file_to_json_string
    **adapter_argument_map
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio/adapters/__init__.py", line 144, in read_from_file
    **adapter_argument_map
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio/adapters/adapter.py", line 136, in read_from_file
    **adapter_argument_map
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio/plugins/python_plugin.py", line 128, in _execute_function
    return (getattr(self.module(), func_name)(**kwargs))
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 908, in read_from_file
    _fix_transitions(result)
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 656, in _fix_transitions
    _fix_transitions(thing.tracks)
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 700, in _fix_transitions
    _fix_transitions(child)
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 692, in _fix_transitions
    child.source_range = child.trimmed_range()
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio/core/item.py", line 98, in trimmed_range
    return self.available_range()
  File "/Users/dev/BitBucket/Project/Submodules/OpenTimelineIO/opentimelineio/schema/clip.py", line 98, in available_range
    self
opentimelineio.exceptions.CannotComputeAvailableRangeError: No available_range set on media reference on clip: Clip("Calvert, Horace_009955R18_10", MissingReference(None, None, {}), None, {'AAF': {'Name': u'Calvert, Horace_009955R18_10', 'SourceID': 'urn:smpte:umid:060a2b34.01010101.01010f00.13000000.060e2b34.7f7f2a80.562848f9.c9e60fd0', 'DataDefinition': {u'Identification': '78e1ebe1-6cef-11d2-807d-006008143e6f', u'Name': u'Sound', u'Description': u'Sound data (legacy)'}, 'ClassName': 'SourceClip', 'Length': 23, 'SourceMobSlotID': 1, 'StartTime': 2524}})
```

Rolling back through the commits, as soon as I hit #462 I can then read the file correctly. Looks like it might be something to do with the change made to `/opentimelineio_contrib/adapters/advanced_authoring_format.py` made in #460 ?

**Comment by freesonluxo** [comment]
I can't quite tell...is this going from otio -> aaf or from aaf -> otio? I think I have seen this before, and it might be worthwhile regenerating the otio file, if it's otio -> aaf.

**Comment by andrewmoore-nz** [comment]
This is reading an AAF generated from Media Composer in to OTIO. I actually had generated an OTIO version of this file a few weeks ago and had it there, which is what prompted me to question it in the first place.

**Comment by freesonluxo** [comment]
For the record, I think I just hit this myself and I'm actively looking into it. Thanks for filing this issue! (Saves me from filing one!)

**Comment by freesonluxo** [comment]
Hey Andrew. Can you try reproing this again with today's master?

If it still is erroring, I have a hunch of what's going on. It looks like that clip doesn't have an available range. #460 addressed incorrect lengths on clips and basically added a check during conversion to only save clip lengths on composition mob clips. That if-check might be restricting it from saving an available_range, which is causing the error.

In any case, could you try reproing again and if it does, perhaps send me an AAF that repros? I can't repro with any AAF file in my collection.

**Comment by andrewmoore-nz** [comment]
Hey @freesonluxo,

Have just tried this again using latest commit and still seeing the same issue. 

> opentimelineio.exceptions.CannotComputeAvailableRangeError: No available_range set on media reference on clip: Clip("Calvert, Horace_009955R18_10", MissingReference(None, None, {}), None, {'AAF': {'Name': u'Calvert, Horace_009955R18_10', 'SourceID': 'urn:smpte:umid:060a2b34.01010101.01010f00.13000000.060e2b34.7f7f2a80.562848f9.c9e60fd0', 'DataDefinition': {u'Identification': '78e1ebe1-6cef-11d2-807d-006008143e6f', u'Name': u'Sound', u'Description': u'Sound data (legacy)'}, 'ClassName': 'SourceClip', 'Length': 23, 'SourceMobSlotID': 1, 'StartTime': 2524}})

Unfortunately I'm unable to send a copy of that AAF (or any other the other ones) where this issue is showing up. Best I can do is give you a copy of the OTIO code that is generated when parsed using the older commit for the same clip:

```
{
    "OTIO_SCHEMA": "Clip.1", 
    "effects": [], 
    "markers": [], 
    "media_reference": {
        "OTIO_SCHEMA": "MissingReference.1", 
        "available_range": null, 
        "metadata": {}, 
        "name": null
    }, 
    "metadata": {
        "AAF": {
            "ClassName": "SourceClip", 
            "DataDefinition": {
                "Description": "Sound data (legacy)", 
                "Identification": "78e1ebe1-6cef-11d2-807d-006008143e6f", 
                "Name": "Sound"
            }, 
            "Length": 23, 
            "Name": "Calvert, Horace_009955R18_10", 
            "SourceID": "urn:smpte:umid:060a2b34.01010101.01010f00.13000000.060e2b34.7f7f2a80.562848f9.c9e60fd0", 
            "SourceMobSlotID": 1, 
            "StartTime": 2524
        }
    }, 
    "name": "Calvert, Horace_009955R18_10", 
    "source_range": {
        "OTIO_SCHEMA": "TimeRange.1", 
        "duration": {
            "OTIO_SCHEMA": "RationalTime.1", 
            "rate": 23.976023976023978, 
            "value": 23
        }, 
        "start_time": {
            "OTIO_SCHEMA": "RationalTime.1", 
            "rate": 23.976023976023978, 
            "value": 88924
        }
    }
}, 
```

Currently, I'm having to roll back to that commit when ever I want to parse any of our more complex AAFs that have come from large. real world projects. I'm not really sure how I would/could go about recreating this error specifically in one of our test project AAFs. However, if that's just simply not enough information to troubleshoot, I'll see what our options are.

**Comment by andrewmoore-nz** [comment]
Can confirm that not a single one of our AAFs from Media Composer can be read by OTIO due to this error.

**Comment by freesonluxo** [comment]
Hey Andrew.

I think I might have an attempt at a fix in my own forked branch. Can you give it a shot and see if it helps at all?

https://github.com/freesonluxo/OpenTimelineIO/tree/otio-482

```
git clone https://github.com/freesonluxo/OpenTimelineIO.git
git checkout otio-482
```

**Comment by andrewmoore-nz** [comment]
Hey @freesonluxo ,

Was able to pull down your branch and test it out, and that's working perfectly - standard OTIO output. 

I was a test on the exact same AAF between your branch the the current master branch commit - yours work as expected, the master gives me this error:

```
WARNING:root:fat sector count missmatch difat: 490 header: 424
Traceback (most recent call last):
  File "/usr/local/bin/otiocat", line 11, in <module>
    load_entry_point('OpenTimelineIO==0.10.0.dev1', 'console_scripts', 'otiocat')()
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio/console/otiocat.py", line 132, in main
    read_adapter_arg_map
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio/console/otiocat.py", line 99, in _otio_compatible_file_to_json_string
    **adapter_argument_map
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio/adapters/__init__.py", line 144, in read_from_file
    **adapter_argument_map
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio/adapters/adapter.py", line 136, in read_from_file
    **adapter_argument_map
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio/plugins/python_plugin.py", line 128, in _execute_function
    return (getattr(self.module(), func_name)(**kwargs))
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 937, in read_from_file
    _fix_transitions(result)
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 685, in _fix_transitions
    _fix_transitions(thing.tracks)
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 729, in _fix_transitions
    _fix_transitions(child)
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio_contrib/adapters/advanced_authoring_format.py", line 721, in _fix_transitions
    child.source_range = child.trimmed_range()
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio/core/item.py", line 98, in trimmed_range
    return self.available_range()
  File "/Users/dev/BitBucket/conformalizer5/Submodules/OpenTimelineIO/opentimelineio/schema/clip.py", line 98, in available_range
    self
opentimelineio.exceptions.CannotComputeAvailableRangeError: No available_range set on media reference on clip: Clip("Grover, Walter Ernest 'Josh'_010441R03_01", MissingReference(None, None, {}), None, {'AAF': {'Name': u"Grover, Walter Ernest 'Josh'_010441R03_01", 'SourceID': 'urn:smpte:umid:060a2b34.01010101.01010f00.13000000.060e2b34.7f7f2a80.5643ebba.88f929f9', 'DataDefinition': {u'Identification': '78e1ebe1-6cef-11d2-807d-006008143e6f', u'Name': u'Sound', u'Description': u'Sound data (legacy)'}, 'ClassName': 'SourceClip', 'Length': 61, 'SourceMobSlotID': 1, 'StartTime': 7058}})
```

**Comment by freesonluxo** [comment]
I think this is fixed. The code has been merged into master. 

I have to say, there's still a little mystery/speculation on why this is happening, so please let us know if it works - or if it doesn't work. Thanks!

**Comment by andrewmoore-nz** [comment]
Great! Just pulled down the latest commit and all seems to be playing nicely. Thanks very much for getting that sorted. Glad that we can now switch back to using the current master branch.
